File: backend/src/main/java/com/kiwi/features/users/data/UsersDataMapper.java
package com.kiwi.features.users.data;

import com.kiwi.common.types.Email;
import com.kiwi.common.types.PositiveOrZeroInteger;

import java.time.LocalDate;

import static com.kiwi.common.utils.FormatUtils.formatDate;

public class UsersDataMapper {

    public static UsersDomain toDomainWithoutPoints(UsersDTO dto) {
        return (dto == null) ? null :
                new UsersDomain(
                        new Email(dto.getEmail()),
                        LocalDate.parse(dto.getRegisterDate())
                );
    }

    public static UsersDTO toUsersDTO(UsersDomain domain) {
        return (domain == null) ? null :
                new UsersDTO(
                        domain.getEmail().value(),
                        formatDate(domain.getRegisterDate())
                );
    }

    public static UsersDomain toDomain(UsersPersistence persistence) {
        if (persistence == null) return null;
        
        UsersDomain domain = new UsersDomain(
                new Email(persistence.getEmail()),
                persistence.getRegisterDate()
        );
        
        domain.setCurrentPoints(new PositiveOrZeroInteger(persistence.getCurrentPoints()));
        domain.setTotalPoints(new PositiveOrZeroInteger(persistence.getTotalPoints()));
        
        return domain;
    }

    public static UsersPersistence toPersistence(UsersDomain domain, String hashedPassword) {
        return (domain == null) ? null :
                UsersPersistence.builder()
                        .email(domain.getEmail().value())
                        .hashedPassword(hashedPassword)
                        .registerDate(domain.getRegisterDate())
                        .currentPoints(domain.getCurrentPoints().value())
                        .totalPoints(domain.getTotalPoints().value())
                        .build();
    }

    public static UsersPersistence toPersistenceWithoutPoints(UsersDTO dto, String hashedPassword) {
        return toPersistence(toDomainWithoutPoints(dto), hashedPassword);
    }

    public static UserPointsDTO toPointsDTO(UsersPersistence persistence) {
        return (persistence == null) ? null :
                UserPointsDTO.builder()
                        .currentPoints(persistence.getCurrentPoints())
                        .totalPoints(persistence.getTotalPoints())
                        .build();
    }

    public static UserPointsDTO toPointsDTO(UsersDomain domain) {
        return (domain == null) ? null :
                UserPointsDTO.builder()
                        .currentPoints(domain.getCurrentPoints().value())
                        .totalPoints(domain.getTotalPoints().value())
                        .build();
    }
}

File: mobile/app/src/main/java/com/bellako/kiwi/features/users/data/UsersDataMapper.kt
package com.bellako.kiwi.features.users.data

import android.os.Build
import androidx.annotation.RequiresApi
import com.bellako.kiwi.common.utils.DateUtils.formatDate
import java.time.LocalDate

object UsersDataMapper {
    @RequiresApi(Build.VERSION_CODES.O)
    fun toDomain(dto: UsersDTO): Result<UsersDomain> {
        val emailResult = Email.of(dto.email)
        return emailResult.fold(
            onSuccess = { validEmail ->
                val passwordResult = Password.of(dto.password)
                passwordResult.fold(
                    onSuccess = { validPassword ->
                        Result.success(UsersDomain(validEmail, validPassword, LocalDate.parse(dto.registerDate)))
                    },
                    onFailure = { err -> Result.failure(err) },
                )
            },
            onFailure = { err -> Result.failure(err) },
        )
    }

    @RequiresApi(Build.VERSION_CODES.O)
    fun toDomain(state: UsersState): Result<UsersDomain> {
        val emailResult = Email.of(state.email)
        return emailResult.fold(
            onSuccess = { validEmail ->
                val passwordResult = Password.of(state.password)
                passwordResult.fold(
                    onSuccess = { validPassword ->
                        Result.success(UsersDomain(validEmail, validPassword, LocalDate.parse(state.registerDate)))
                    },
                    onFailure = { err -> Result.failure(err) },
                )
            },
            onFailure = { err -> Result.failure(err) },
        )
    }

    @RequiresApi(Build.VERSION_CODES.O)
    fun toState(domain: UsersDomain): UsersState =
        UsersState(
            email = domain.email.value,
            password = domain.password.value,
            registerDate = formatDate(domain.registerDate),
        )

    @RequiresApi(Build.VERSION_CODES.O)
    fun toState(dto: UsersDTO): Result<UsersState> {
        val domain = toDomain(dto)
        return domain.fold(
            onSuccess = { validDomain ->
                Result.success(toState(validDomain))
            },
            onFailure = { err -> Result.failure(err) },
        )
    }

    @RequiresApi(Build.VERSION_CODES.O)
    fun toDTO(domain: UsersDomain): UsersDTO =
        UsersDTO(
            email = domain.email.value,
            password = domain.password.value,
            registerDate = formatDate(domain.registerDate),
        )

    @RequiresApi(Build.VERSION_CODES.O)
    fun toDTO(state: UsersState): Result<UsersDTO> {
        val domain = toDomain(state)
        return domain.fold(
            onSuccess = { validDomain ->
                Result.success(toDTO(validDomain))
            },
            onFailure = { err -> Result.failure(err) },
        )
    }
}
